### PR TITLE
PlutusLedgerApi.V2 support; Inline Datum, Reference Inputs, and more

### LINKED ISSUE #15: Optional Phase-1 Validation

### PR DESCRIPTION
**Pull request by SeungheonOh**
close #15 and #18

### CODE REVIEW

File: src/Plutarch/Context/Base.hs [hunk 1]
```diff
@@ -310,59 +322,63 @@ output ::
...
-  where
-    g x = mempty{bbInputs = pure x}
+input x = pack mempty{bbInputs = pure x}
+
+reference ::
+    forall (a :: Type).
+    (Builder a) =>
+    UTXO ->
+    a
+reference x = pack mempty{bbReferenceInputs = pure x}
```
Comment (peter-mlabs) [inline src/Plutarch/Context/Base.hs:339]: This name should be more descriptive -- CIP 31 is "reference inputs", CIP 33 is "reference scripts". How does `referenceInput` sound?

File: src/Plutarch/Context/Phase1.hs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+checkFoldable :: Foldable t => Checker a -> Checker (t a)
+checkFoldable c = Checker $ \y -> foldMap (runChecker c) y
+
+-- | @since 2.1.0
+checkIf :: (a -> Bool) -> CheckerError -> Checker a
+checkIf f err = Checker $ \y -> if f y then mempty else pure err
+
+-- | @since 2.1.0
+checkByteString :: Checker BuiltinByteString
+checkByteString = Checker $ \y -> if lengthOfByteString y == 28 then mempty else pure $ IncorrectByteString
```
Comment (peter-mlabs) [inline src/Plutarch/Context/Phase1.hs:80]: Could you explain this one?

File: src/Plutarch/Context/Phase1.hs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+-- | @since 2.1.0
+checkValue :: Checker Value
+checkValue = Checker $ \y -> if isPos y then mempty else pure $ NonPositiveValue y
+  where
+    isPos = all (\(_, _, x) -> x > 0) . flattenValue
```
Comment (peter-mlabs) [inline src/Plutarch/Context/Phase1.hs:86]: This needs a better name and documentation. 

Some notes:

- `Value` can contain negative entries if it's part of a mint (negative entries represent burning tokens)
- `Value` can contain a zero entry if the asset class is `(adaSymbol, adaToken)`, and _must_ contain a zero entry if the value does not contain Ada.

File: src/Plutarch/Context/Phase1.hs (hunk 2 above)
Comment (peter-mlabs) [reply src/Plutarch/Context/Phase1.hs:80]: Oh, I see now this is for checking `TxId`s and `PubKeyHash`es. We should add a doc string here.

File: src/Plutarch/Context/Phase1.hs [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+ @since 2.1.0
+-}
+checkSignatures :: Builder a => Checker a
+checkSignatures =
+    mconcat $
+        [ contramap (fmap getPubKeyHash . bbSignatures . unpack) (checkFoldable checkByteString)
+        , contramap (length . bbSignatures . unpack) (checkIf (>= 1) NoSignature)
+        ]
+
+{- | Check if input, output, mint have zero sum.
```
Comment (peter-mlabs) [inline src/Plutarch/Context/Phase1.hs:106]: Does the fee need to be included in this too?

Comment (peter-mlabs) [reply src/Plutarch/Context/Phase1.hs:106]: Oh, I see that its in the implementation. Lets add it to the doc string.

**Review by peter-mlabs — CHANGES_REQUESTED** [review]
Some changes, mostly needing to add some more doc strings.

`Phase1` is a good start to a complicated problem, but I think it needs some renaming and an overall module doc string indicating that there is a very small subset of the total checks that we want. After this is merged, I wouldn't put too much more effort into this until we've worked something out with Jared.

**Comment by SeungheonOh** [comment]
Some sample of new error messages:
```
    Error at Input:
        Given validator input does not exist in inputs: 
            ValidatorOutRefIdx 10
```
```
    Error at Input:
        Value (Map [(,Map [("",100)]),(cc,Map [("hello",-50)])]) is an invalid value (contains non positive).
    Error at Fee:
        Transaction takes tokens other than Ada as fee: Value (Map [(cc,Map [("non Ada Fee",330)])])
    Error at TxInfo:
        Transaction doesn't not have equal inflow and outflow.
        Diff: Value (Map [(aaaa,Map [("hello",-3)]),(cc,Map [("hello",-100),("non Ada Fee",330)])])
    Error at Signatories:
        Signature not provided
```

File: sample/Main.hs [hunk 5]
```diff
@@ -1,39 +1,49 @@
+{-# OPTIONS_GHC -Wno-all #-}
+{-# OPTIONS_GHC -Wno-unused-imports #-}
```
Comment (kozross) [inline sample/Main.hs:2]: This shouldn't be here.

File: src/Plutarch/Context/Base.hs [hunk 6]
```diff
@@ -403,11 +415,39 @@ yieldInInfoDatums (toList -> inputs) (unpack -> bb)
...
+ incremental indices to each inputs starting from one. If there
+ are duplicate reserved indices, the second occurance will be
+ treated as non-reserved and given incremental index.
+
+ @since 2.1.0
+-}
+mkOutRefIndices ::
+    Builder a =>
+    a ->
+    a
```
Comment (kozross) [inline src/Plutarch/Context/Base.hs:438]: Kind signature.

File: src/Plutarch/Context/SubBuilder.hs [hunk 7]
```diff
@@ -27,7 +37,11 @@ import PlutusLedgerApi.V1.Scripts (
 -}
 newtype SubBuilder
     = SubBuilder BaseBuilder
-    deriving newtype (Semigroup, Monoid, Builder)
+    deriving newtype (Semigroup, Monoid)
```
Comment (kozross) [inline src/Plutarch/Context/SubBuilder.hs:40]: Not a problem per se, but I prefer `deriving via` even for newtypes. I've been bitten several times by changes in underlying representation silently changing instance behaviour without the compiler reminding me.

File: src/Plutarch/Context/SubBuilder.hs [hunk 8]
```diff
@@ -57,16 +71,15 @@ buildTxOuts (unpack -> BB{..}) = utxoToTxOut <$> toList bbOutputs
 
  @since 2.0.0
 -}
-buildTxInInfos :: SubBuilder -> Either String [TxInInfo]
-buildTxInInfos b@(unpack -> BB{..}) = flip runContT Right $ do
-    (ins, _) <- yieldInInfoDatums bbInputs b
-    return ins
+buildTxInInfos :: SubBuilder -> [TxInInfo]
+buildTxInInfos (unpack -> BB{..}) =
+    fst $ yieldInInfoDatums bbInputs
```
Comment (kozross) [inline src/Plutarch/Context/SubBuilder.hs:76]: Changes in signature require a change in their `since`

**Review by kozross — CHANGES_REQUESTED** [review]
A few minor things, mostly to do with `since`s and changelog entries.

**Comment by peter-mlabs** [comment]
I think this can be g2g once the failing test is fixed.

**Comment by SeungheonOh** [comment]
Ah okay, I thought I fixed it already. I'll fix it really quick.

**Pull request merged**